### PR TITLE
ByteArray# literals

### PR DESCRIPTION
**Pull request by phadej**
This is a proposal to introduce `ByteArray#` and `(# Word#, Addr# #)` literals, and slightly change `Addr#` literals. In short you'll be able to write

```haskell
"Literals"#b           -- ByteArray#
"\xef\xbb\xbf"#abytes  -- Addr#
"Юникод"#ucp1251       -- (# Int#, Addr# #)
```

[Rendered](https://github.com/phadej/ghc-proposals/blob/bytearray-literals/proposals/0000-bytearray-literals.rst)

### CODE REVIEW

**Comment by cartazio** [comment]
I like this a lot

File: proposals/0000-bytearray-literals.rst [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+This would be a potentially breaking change for any code using unboxed string literals.
+
+Alternatives
+------------
+
+One proposed alternative is to use double hash syntax ``"foo"##`` to represent
+UTF8 ``ByteArray#``.  That variant is very limited in power compared to proposed.
+
+In fact, we might introduce similar syntax for the number literals, e.g. ``120#i8``, if we get `Int8#` primitive type.
```
Comment (bgamari) [inline proposals/0000-bytearray-literals.rst]: I agree that the syntax that you propose above is *much* preferred to the "keep-adding-more-hashes" approach. I would really like to see the this syntax adopted for numeric llterals at some point as well.

File: proposals/0000-bytearray-literals.rst [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+will mean the same as
+
+.. code-block:: haskell
+
+  "hello"#aLatin1
+
+Therefore, code with current primitive strings won't break.
+
+*Unresolved:* Should there be a (``-Wall``) warning in this case, asking user to be explicit?
```
Comment (bgamari) [inline proposals/0000-bytearray-literals.rst]: I'm not sure I see the advantage. Implicit is fine given this syntax has a long history here and avoids the nasty compatibility issues that warning generally brings.

File: proposals/0000-bytearray-literals.rst (hunk 1 above)
Comment (cartazio) [reply proposals/0000-bytearray-literals.rst]: that'd be nice, agreed

File: proposals/0000-bytearray-literals.rst [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+  "Literals"#b           -- ByteArray#
+  "\xef\xbb\xbf"#abytes  -- Addr#
+  "Юникод"#ucp1251       -- (# Int#, Addr# #)
+
+additionally to current
+
+.. code-block:: haskell
+
+  "primitive"#           -- Addr# in utf8
+  "string                -- String or IsString a => a
```
Comment (nickkuk) [inline proposals/0000-bytearray-literals.rst]: No quotes?

File: proposals/0000-bytearray-literals.rst [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+
+Primitive string without modifier
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The current primitive string
+
+.. code-block:: haskell
+
+  "hello#
```
Comment (nickkuk) [inline proposals/0000-bytearray-literals.rst]: No quotes?

File: proposals/0000-bytearray-literals.rst [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+Currently, it's possible to create primitive ``Addr#`` string literals:
+
+.. code-block:: haskell
+
+  "hello"# -- :: Addr#
+
+These literals are ``[Word8]`` literals, *primitive string literal must contain only characters <= '\xFF'*.
+
+Ordinary strings, like ``"hello``, ``"Юникод"``, ``"\NUL"`` are then desugared as
```
Comment (nickkuk) [inline proposals/0000-bytearray-literals.rst]: No quotes after "hello".

File: proposals/0000-bytearray-literals.rst [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/134>`_.
```
Comment (nickkuk) [inline proposals/0000-bytearray-literals.rst]: 134 should be 135.

**Comment by cartazio** [comment]
Question:
How will escaping be handled? Or is that already addressed via the proposal
implicitly?

File: proposals/0000-bytearray-literals.rst (hunk 3 above)
Comment (phadej) [reply proposals/0000-bytearray-literals.rst]: Thanks for spotting. Correct others as well.

**Comment by simonpj** [comment]
The proposal helpfully lists a bunch of motivations with Trac tickets.  I like that it resolves a bunch of related issues.  But could you add a section going through these tickets one by one and explaining how the tickets are thereby resolved?

Trac #5218 speaks of
```
{-# LANGUAGE OverloadedStrings #-}
module Foo where
import Data.ByteString.Char8
foo = "abc" :: ByteString
```
How will that desugar.  Will good things happen?

**Comment by simonmar** [comment]
The motivation has the example

```
"Юникод"#ucp1251       -- (# Int#, Addr# #)
```

but `cp1251` isn't one of the allowed encodings according to the proposal.

What does "algebraic" mean in the table in the motivation section? 

Under "string syntax desugaring", the proposal says we'll use `ByteArray#`, but I'm a bit worried about the potential increase in code size here. The proposal refers to https://ghc.haskell.org/trac/ghc/ticket/5218#comment:95 for evidence that the overhead is small, but I think we should re-check that: it only considers the addition of a single word per string literal, but we would need 2 (the header + size).

Why not use `(# Addr#, Int# #)` for string literals, as we've discussed before? The proposal says "Desugaring to ByteArray# literals will allow GHC to eliminate common (sub)-expressions.", is this referring to the problem in https://ghc.haskell.org/trac/ghc/ticket/11312? We need to be really careful here - `ByteArray#` could have the same problem, if the compiler can inline and duplicate a `ByteArray#` literal.  Anyway, I think we need to consider this point very carefully, and the proposal should elaborate on it.

Would these `ByteArray#`s return `True` for `isByteArrayPinned#`? (presumably yes)

The proposal doesn't really motivate the addition of the `(# Addr#, Int# #)` version. What would it be used for, if not string literals?

**Comment by nomeata** [comment]
There is some good feedpack. @phadej, are you going to incoporate it into the proposal?

**Comment by phadej** [comment]
I will. A bit busy with other stuff right now.

**Comment by fumieval** [comment]
Any updates? A lot of people shot themselves in their feet due to `IsString ByteString` with multi-byte strings. This proposal seems like a good solution.

File: proposals/0000-bytearray-literals.rst [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+-----------------------
+
+Library authors may provide `RULES` to avoid looping to get the length of the string (``Prim.Builder``  in ``bytestring`` could simply ``memcpy`` the contents of a literal),
+or avoid roundtripping through `String` (also in ``bytestring``):
+
+.. code-block:: haskell
+
+  data ShortByteString = SBS ByteArray#
+ 
+  instance FromString ShortByteString where
```
Comment (fumieval) [inline proposals/0000-bytearray-literals.rst]: I think you are talking about the `IsString` class.

**Comment by phadej** [comment]
@fumieval I might look into updating the proposal during holidays

**Comment by hsyl20** [comment]
It seems to me that the proposed literals are just TH expression quasi-quoters with some very thin syntactic sugar:
```haskell
 "abc123"#enc <=> [enc|abc123|]
```
So i think it would be better to reuse and to extend quasi-quoters instead:

1) Add more built-in quasi-quoters

Currently GHC (via TH) supports "e", "t", "d" and "p" quasi-quoters. Add some common ones, for instance:

QuasiQuoter | Input | Expr Type
-----------------|----------|-----
 cstring | any string | `Addr# {- Null terminated UTF-8 encoded string-}`
 bytes | ASCII string | `(# Addr#, Word# #)`
hexbin | [0-9A-F]* | `(# Addr#, Word# #)`
utf8 | any string | `(# Addr#, Word# {- Size in bytes -}, Word# {- Number of code-points #-} #)`
utf16le | any string | `(# Addr#, Word# {- Size in bytes -}, Word# {- Number of code-points #-} #)`
... | ... | ...

As usual we can export names from ghc-prim for built-in quasi-quoters in order to require an "import" to make those quasi-quoters in scope instead of polluting the global namespace.

2) Make it extensible: allow GHC plugins to define and register "builtin" quasi-quoters.

Built-in or plugged-in quasi-quoters don't need template-haskell (which is not always available).

3) (Optional) Add syntactic sugar: `"xyz"#enc ==> [enc|xyz|]"`

4) Make it more efficient

Redefine `QuasiQuoter` to be:

```
data QuasiQuoter = QuasiQuoter
   { -- For backward compatibility
     quoteExp :: String -> Q Exp
   , quotePat :: String -> Q Pat
   , quoteType :: String -> Q Type
   , quoteDec :: String -> Q [Dec]
   -- Operations on raw UTF-8 buffers
   , quoteExpRaw :: Addr# -> Word# -> Q Exp
   , quoteExpRaw a sz = quoteExp (unpackNBytes# a (word2Int# sz))
   , quotePatRaw :: Addr# -> Word# -> Q Pat
   , quotePatRaw a sz = quotePat (unpackNBytes# a (word2Int# sz))
   , quoteTypeRaw :: Addr# -> Word# -> Q Type
   , quoteTypeRaw a sz = quoteType (unpackNBytes# a (word2Int# sz))
   , quoteDecRaw :: Addr# -> Word# -> Q [Dec]
   , quoteDecRaw a sz = quoteDec (unpackNBytes# a (word2Int# sz))
   }
```

and use the "raw" operations.

**Comment by phadej** [comment]
@hsyl20 AFAIK QuasiQuotes cannot be used with stage1 compiler, i.e. you cannot use them in `bytestring`

Also this proposal is not only about syntax.

**Comment by hsyl20** [comment]
@phadej yes I know. See point (2): "Built-in or plugged-in quasi-quoters don't need template-haskell (which is not always available)."

The idea is to extract quasi-quoters from TH to put them into GHC in order to generalize them and to reuse their syntax:
* built-in GHC quasi-quoters: cstring, utf8, bytes, whatever (as in your proposal)
* when TH is available: e,p,t,d
* extensible via GHC plugins: because inevitably someone will require an encoding that isn't built-in into GHC
* when TH is available: extensible via TH as currently

**Comment by phadej** [comment]
@hsyl20 

> Make it extensible: allow GHC plugins to define and register "builtin" quasi-quoters.

Is worth own proposal. I won't even try to sneak it in through this one.

**Comment by phadej** [comment]
I won't be able to finish this proposal in a foreseeable future

**Pull request closed without merging**